histogram/src/config.rs

```diff
@@ -0,0 +1,237 @@
+//! The configuration of a histogram which determines the buckets and how to
+//! convert a value to a bucket index and vice versa.
+
+use super::{BuildError, Error};
+use crate::RangeInclusive;
+
+#[derive(Clone, Copy)]
+pub(crate) struct Config {
+    max: u64,
+    a: u8,
+    b: u8,
+    n: u8,
+    cutoff_power: u8,
+    cutoff_value: u64,
+    lower_bin_count: u32,
+    upper_bin_divisions: u32,
+    upper_bin_count: u32,
+}
+
+impl Config {
+    pub fn new(a: u8, b: u8, n: u8) -> Result<Self, BuildError> {
+        // temporarily convert these to wider types
+        let a: u32 = a.into();
+        let b: u32 = b.into();
+        let n: u32 = n.into();
+
+        // we only allow values up to 2^64
+        if n > 64 {
+            return Err(BuildError::MaxPowerTooHigh);
+        }
+
+        // check that the other parameters make sense together
+        if a + b >= n {
+            return Err(BuildError::MaxPowerTooLow);
+        }
+
+        // the cutoff is the point at which the linear range divisions and the
+        // logarithmic range subdivisions diverge.
+        //
+        // for example:
+        // when a = 0, the linear range has bins with width 1.
+        // if b = 7 the logarithmic range has 128 subdivisions.
+        // this means that for 0..128 we must be representing the values exactly
+        // but we also represent 128..256 exactly since the subdivisions divide
```

Comment:
I don't get this "subdivision" reasoning. 

We have the linear portion being twice as wide purely for bin lookup efficiency reasons, because the sum of a power-of-2 geometrical sequence ends up being `2^n - 1`

Reply:
It's maybe not the best description. I think your comment about rephrasing this to avoid the distinction between the 'linear' and 'logarithmic' ranges can help us to clarify this language.